tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,518 @@
+    graph: A tf.Graph instance for the original graph.
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+    previously_visited: A set of node names from previous subgraph layers.
+    output_node_names: A list of output node names for the current subgraph.
+
+  Returns:
+    An ExecutionSpec representing a subgraph layer.
+  """
+  subgraph = graph_pb2.GraphDef()
+  subgraph.versions.CopyFrom(graph_def.versions)
+  subgraph.library.CopyFrom(graph_def.library)
+
+  queue = collections.deque(output_node_names)
+  visited = set()
+
+  while queue:
+    current_node_name = queue.popleft()
+    current_node = node_name_to_node_def[current_node_name]
+
+    if current_node_name not in visited:
+      visited.add(current_node_name)
+
+      if _is_remote_op(current_node) or _is_placeholder_op(current_node) \
+      or current_node_name in previously_visited:
+        # These ops must be computed before this subgraph layer. Hence,
+        # we treat them as placeholder inputs.
+        placeholder_node = _create_placeholder_node_from_existing_node(
+            current_node, graph)
+        subgraph.node.append(placeholder_node)
+
+      else:
+        subgraph.node.append(current_node)
+        queue.extend(node_name_to_node_def[current_node_name].input)
+
+  return ExecutionSpec(
+      subgraph=subgraph,
+      input_names=_get_input_names(subgraph),
+      output_names=output_node_names,
+      is_remote_op=False)
+
+
+def _create_placeholder_node_from_existing_node(
+    node: node_def_pb2.NodeDef, graph: tf.Graph) -> node_def_pb2.NodeDef:
+  """Creates a placeholder node to represent an existing node.
+
+  Some partitioned subgraphs may require inputs that are loaded or computed
+  previously. Hence, we replace the input nodes with placeholder nodes that
+  share the same name, shape, and dtype. Now the inputs become placeholders
+  inside partitioned subgraphs, and can be loaded by feed dicts at the runtime.
+
+  Args:
+    node: A `NodeDef` proto for the existing node.
+    graph: A tf.Graph instance for the graph that contains the existing node.
+
+  Returns:
+    A `NodeDef` proto that stores a placeholder node.
+  """
+  operation = graph.get_operation_by_name('import/%s' % (node.name))
+  output_tensor = operation.outputs[0]
+
+  with tf.compat.v1.Session(graph=tf.Graph()) as sess:
+    tf.compat.v1.placeholder(dtype=output_tensor.dtype,
+                             shape=output_tensor.shape,
+                             name=node.name)
+    return sess.graph_def.node[0]
+
+
+def _get_input_names(subgraph: graph_pb2.GraphDef) -> List[Text]:
+  input_names = [node.name for node in subgraph.node
+                 if _is_placeholder_op(node)]
+  return input_names
+
+
+def _get_non_input_names(subgraph: graph_pb2.GraphDef) -> List[Text]:
+  non_input_names = [node.name for node in subgraph.node
+                     if not _is_placeholder_op(node)]
+  return non_input_names
+
+
+def _partition_one_remote_op_layer(
```

Comment:
Don't name this `partition` - it doesn't do any partitioning.

`_remote_op_layer_to_execution_specs`? `get_execution_specs_for_remote_op_layer`?

Reply:
Addressed.